File: Connect_4_Using_C/game.c
#include "game.h"

void initializeGame(GameState *game) {
    for (int i = 0; i < ROWS; i++)
        for (int j = 0; j < COLS; j++)
            game->board[i][j] = EMPTY;

    game->currentPlayer = PLAYER;
    game->gameOver = false;
    game->winner = EMPTY;
}

bool isValidMove(GameState *game, int col) {
    return col >= 0 && col < COLS && game->board[0][col] == EMPTY;
}

bool makeMove(GameState *game, int col) {
    if (!isValidMove(game, col)) return false;

    int row;
    for (row = ROWS - 1; row >= 0; row--)
        if (game->board[row][col] == EMPTY) break;

    game->board[row][col] = game->currentPlayer;

    if (checkWin(game, game->currentPlayer)) {
        game->gameOver = true;
        game->winner = game->currentPlayer;
    } else if (checkDraw(game)) {
        game->gameOver = true;
        game->winner = EMPTY;
    }

    game->currentPlayer = (game->currentPlayer == PLAYER) ? AI : PLAYER;
    return true;
}

bool checkDraw(GameState *game) {
    for (int col = 0; col < COLS; col++)
        if (game->board[0][col] == EMPTY) return false;
    return true;
}

bool checkWin(GameState *game, int player) {
    for (int row = 0; row < ROWS; row++)
        for (int col = 0; col <= COLS - 4; col++)
            if (game->board[row][col] == player &&
                game->board[row][col+1] == player &&
                game->board[row][col+2] == player &&
                game->board[row][col+3] == player) return true;

    for (int row = 0; row <= ROWS - 4; row++)
        for (int col = 0; col < COLS; col++)
            if (game->board[row][col] == player &&
                game->board[row+1][col] == player &&
                game->board[row+2][col] == player &&
                game->board[row+3][col] == player) return true;

    for (int row = 0; row <= ROWS - 4; row++)
        for (int col = 0; col <= COLS - 4; col++)
            if (game->board[row][col] == player &&
                game->board[row+1][col+1] == player &&
                game->board[row+2][col+2] == player &&
                game->board[row+3][col+3] == player) return true;

    for (int row = 3; row < ROWS; row++)
        for (int col = 0; col <= COLS - 4; col++)
            if (game->board[row][col] == player &&
                game->board[row-1][col+1] == player &&
                game->board[row-2][col+2] == player &&
                game->board[row-3][col+3] == player) return true;

    return false;
}

File: model/Game.java
package com.game.connect4.model;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;


public class Game {
    private String id;
    private Board board;
    private GameStatus status;
    private Player currentPlayer;
    private int difficultyLevel;
    private List<String> gameAnalysis;

    public Game(int difficultyLevel) {
        this.id = UUID.randomUUID().toString();
        this.board = new Board();
        this.status = GameStatus.IN_PROGRESS;
        this.currentPlayer = Player.PLAYER; // Player starts first
        this.difficultyLevel = difficultyLevel;
        this.gameAnalysis = new ArrayList<>();
    }


    public void resetGame() {
        this.board = new Board();
        this.status = GameStatus.IN_PROGRESS;
        this.currentPlayer = Player.PLAYER;
        this.gameAnalysis.clear();
    }


    public int makeMove(int column) {
        if (status != GameStatus.IN_PROGRESS) {
            return -1;
        }

        int row = board.makeMove(column, currentPlayer);
        if (row == -1) {
            return -1;
        }

        // Check for win
        if (board.checkWin(row, column, currentPlayer)) {
            status = (currentPlayer == Player.PLAYER) ? GameStatus.PLAYER_WIN : GameStatus.AI_WIN;
        }
        // Check for draw
        else if (board.isFull()) {
            status = GameStatus.DRAW;
        }

        // Switch players if game is still in progress
        if (status == GameStatus.IN_PROGRESS) {
            currentPlayer = currentPlayer.getOpponent();
        }

        return row;
    }

    public String getId() {
        return id;
    }

    public Board getBoard() {
        return board;
    }

    public GameStatus getStatus() {
        return status;
    }

    public Player getCurrentPlayer() {
        return currentPlayer;
    }

    public int getDifficultyLevel() {
        return difficultyLevel;
    }

    public void setDifficultyLevel(int difficultyLevel) {
        this.difficultyLevel = difficultyLevel;
    }

    public List<String> getGameAnalysis() {
        return gameAnalysis;
    }

    public void addAnalysis(String analysis) {
        this.gameAnalysis.add(analysis);
    }


    public Game copy() {
        Game gameCopy = new Game(this.difficultyLevel);
        gameCopy.board = new Board(this.board);
        gameCopy.status = this.status;
        gameCopy.currentPlayer = this.currentPlayer;
        return gameCopy;
    }
}